--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -43,9 +43,7 @@ always:
 
 # Always rebuild because .git/HEAD is a symbolic ref one can't depend on
 main/StellarCoreVersion.cpp: always
-	@vers=$$(cd "$(srcdir)" \
-		&& git describe --always --dirty --tags 2>/dev/null \
-		|| echo "$(PACKAGE) $(VERSION)"); \
+	@vers=$$(echo "${LOCAL_PACKAGE} ${LOCAL_VERSION} (${GIT_COMMIT})"); \
 		sed -e "s@%%VERSION%%@$$vers@" \
 			< "$(srcdir)/main/StellarCoreVersion.cpp.in" > $@~
 	@if cmp -s $@~ $@; then rm -f $@~; else \
